### PR TITLE
Do not double close reader

### LINKED ISSUE #29346: Edit file times out for large files after recent cat-file bugfix
### Description

To reproduce:
* Create new new repository with Apache 2.0 license
* Click LICENSE file and Edit File
* Operation times out

Happens for any larger text file, works fine for a small file.

I bisected the cause down to #29297.

### Gitea Version

1.21.6

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Running locally on Linux with own build, but also happens on try.gitea.io.

### Database

None

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #29346

### CODE REVIEW

File: modules/git/blob_nogogit.go [hunk 1]
```diff
@@ -102,7 +103,17 @@ func (b *blobReader) Read(p []byte) (n int, err error) {
 
 // Close implements io.Closer
 func (b *blobReader) Close() error {
+	if b.rd == nil {
+		return fs.ErrClosed
+	}
```
Comment (KN4CK3R) [inline modules/git/blob_nogogit.go]: Should it `panic` here instead? Calling `Close` multiple times is a coding error. I checked all calls to `DataAsync` and could not find another double close. A `panic` could prevent double closes in future while returning a (most of time) ignored error could hide it.

Comment (wxiaoguang) [reply modules/git/blob_nogogit.go]: No. None of Golang's other "Close" methods panics.

Comment (wxiaoguang) [reply modules/git/blob_nogogit.go]: Could it simply return "nil"? Most (all) errors for a Close method are meanless .... no caller cares about it.

**Review by wxiaoguang — APPROVED** [review]
Prefer to return "nil". But could LGTM ahead.

File: modules/git/blob_nogogit.go (hunk 1 above)
Comment (lunny) [reply modules/git/blob_nogogit.go]: I think double Close should be supported. i.e. It's safe to invoke `file.Close` twice. So I think we should also allow invoking `Close` twice. It's convenient for some situations.

**Review by lunny — APPROVED** [review]
`return nil` is better.

File: modules/git/blob_nogogit.go (hunk 1 above)
Comment (KN4CK3R) [reply modules/git/blob_nogogit.go]: Should we then add this rule to our coding guidelines?
> If a type implements `io.Closer`, calling `Close` multiple times must not fail and return a `nil` error.

**Review by wxiaoguang — APPROVED** [review]

File: modules/git/blob_nogogit.go (hunk 1 above)
Comment (wxiaoguang) [reply modules/git/blob_nogogit.go]: I guess it depends, for most cases: yes, `nil` is good enough.

But I guess there could be some special cases that the caller might be interested in the result.

Comment (KN4CK3R) [reply modules/git/blob_nogogit.go]: Then maybe relax the `nil` but specify that calling `Close` multiple times must not be undefined behaviour and must not panic?

**Pull request merged**